### DISCUSSION TITLE
Working with Ardalis.SharedKernel EntityBase<TId>

### DISCUSSION THREAD

**Discussion by ardalis** [SteveDunn/Vogen#677] [General]
Hey Steve,

I'm trying to use Vogen for strongly typed IDs in my projects, many of which leverage something like Ardalis.SharedKernel:
https://github.com/ardalis/ardalis.sharedkernel

Specifically there's base entity class that a lot of other types like repositories and specifications depend on, which looks like this:

```csharp
public abstract class EntityBase : HasDomainEventsBase
{
  public int Id { get; set; }
}

public abstract class EntityBase<TId> : HasDomainEventsBase
  where TId : struct, IEquatable<TId>
{
  public TId Id { get; set; }
}
```

Ok so if I just use this as-is, I get this exception:

```
Message: 
  Vogen.ValueObjectValidationException : Use of uninitialized Value Object at: 

Stack Trace: 
  ThrowHelper.ThrowWhenNotInitialized(StackTrace stackTrace) line 629
  ToDoItemId.GetHashCode() line 489
```

It doesn't like that I never set the value. I looked at your docs and it seems to suggest I use a nullable type for this scenario, but that would break a lot of things. I could specify something like an instance value:

```csharp
[ValueObject<int>]
[Instance("New", 0)]
public partial struct ToDoItemId; 
```

But I don't see how to use this generically from BaseEntity<TId> and if I override the Id property using `new` then other things break.

Any thoughts on how to support Vogen Id types in a generic way following the convention EF Core uses that value types that match their defaults represent net new entities and thus should be treated as New by the change tracker?

Thanks,
Ardalis

**Comment by ardalis** [comment]
Also I tried updating the base class to explicitly set the Id to `default!` but that didn't work, either.

**Comment by ardalis** [comment]
Glad to hear about my other question - any thoughts on this one? I'd prefer to avoid having to deal with nullable Ids everywhere, and it's only ever an issue on initial creation.

**Comment by SteveDunn** [comment]
Hi,
I'm no expert on EFCore, so please excuse the state of the example that I've attached. The example is a simple EFCore app that uses a SQLite file. It's basically the 'blog posts' tutorial, but with Vogen-generated IDs.

The way I got this working was:

```c#
[assembly: VogenDefaults(
    staticAbstractsGeneration: StaticAbstractsGeneration.MostCommon | StaticAbstractsGeneration.InstanceMethodsAndProperties, 
    conversions: Conversions.EfCoreValueConverter)]
```

This tells Vogen to generate static abstract interfaces and, along with this, generate the interface declarations for the `From` method. Given that, all value objects generated now have an `IVogen` interface:

```csharp
public partial struct BlogId : ...IVogen<BlogId, int>
```

Given that, we can put some generic constraints on a value generator:
```csharp
internal class VogenIdValueGenerator<TContext, TEntityBase, TId> : ValueGenerator<TId>
    where TContext : DbContext
    where TEntityBase : EntityBase<TEntityBase, TId> 
    where TId : IVogen<TId, int>
```

However, I did have to change `EntityBase` slightly:
```
public abstract class EntityBase<T, TId> where T: EntityBase<T, TId>
```

We can the do:
```csharp
        modelBuilder.Entity<Blog>(b =>
        {
            b.HasKey(x => x.Id);
            b.Property(e => e.Id)
                .HasValueGenerator<VogenIdValueGenerator<BloggingContext, Blog, BlogId>>()
                .HasVogenConversion();
        });
```

There may be other ways of doing this, but I hope this helps. Please do let me know how you get on if there's anything else I can help with.


[efcore-and-vogen.zip](https://github.com/user-attachments/files/17216577/efcore-and-vogen.zip)

**Reply by ardalis** [reply]
Thanks! I'm back from a trip and getting back into this. I'd love to have a way to avoid having EF-core specific stuff involved in the project where the base class is defined. It should be "persistence ignorant". I'm fine to take a dependency on Vogen if needed, of course. Is there a way if I have a low level shared project with the base type (Ardalis.SharedKernel in this case) for it to have the necessary Vogen-specific attributes, types, etc. but *not* any EF Core stuff, and then in some consuming project (for instance, Ardalis.CleanArchitecture.Infrastructure) those types would exist.

Looking at the sample it appears the assembly directive that is needed to generate the EF Core Value Converters would need to be in the same project as the model types (entities and their value objects). And I'm trying to keep that project free from EF references...

Putting the VogenIdValueGenerator in my Infrastructure project results in it not being able to find the IVogen interface:
![image](https://github.com/user-attachments/assets/d3a634a1-f859-441d-92a1-e8f92249846c)

So, I think what you have here will work, it's just a matter of me seeing if I can split it up across three projects (Shared Kernel, Core, Infrastructure) such that Infrastructure references Core and Core references Shared Kernel. Then:

EntityBase<T,TId> lives in SharedKernel.

Model types (Blog, Post, etc) live in Core.

EF Core references all live in Infrastructure.

See: https://github.com/ardalis/CleanArchitecture/tree/main/sample

and for my work in progress see branch ardalis/sample-vogen where I'm trying to get the sample working with Vogen VOs for IDs and certain other types.

**Reply by ardalis** [reply]
When I move the assembly attribute to my Core project the IVogen interface is generated but the generated code blows up because it's missing EF Core references:
![image](https://github.com/user-attachments/assets/1ac66be2-2bee-4266-bfbd-d1968c73168e)

**Comment by SteveDunn** [comment]
Hi - I forked and created a PR for your branch. I think it does everything you want. Please do let me know if it doesn't or if you have any ideas on how to make it more 'elegant' (I don't think it'll ever be elegant, but it could probably cause less friction!)

Also, re. the `New` attribute: in hindsight, this attribute was a mistake. It's far better to just create a static field instead: `public static TodoItemId New = new(0);` (the analyser that prohibits new expressions skips the type itself)

**Comment by ardalis** [comment]
I updated the public Ardalis.SharedKernel so it includes the base type needed. I added ProjectId and ToDoItemId value objects and configured both of them to use VogenIdValueGenerator<T,T,T> but I still see the same error in my tests.

I've pushed my latest changes to my branch here:
https://github.com/ardalis/CleanArchitecture/tree/ardalis/sample-vogen/sample


```
NimblePros.SampleToDo.IntegrationTests.Data.EfRepositoryAdd.AddsProjectAndSetsId
   Source: EfRepositoryAdd.cs line 9
   Duration: 13 ms

  Message: 
Vogen.ValueObjectValidationException : Use of uninitialized Value Object at: 

  Stack Trace: 
ThrowHelper.ThrowWhenNotInitialized(StackTrace stackTrace) line 625
ProjectId.GetHashCode() line 487
NoNullsCustomEqualityComparer.GetHashCode(TKey obj)
Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
Dictionary`2.set_Item(TKey key, TValue value)
IdentityMap`1.Add(TKey key, InternalEntityEntry entry, Boolean updateDuplicate)
IdentityMap`1.Add(TKey key, InternalEntityEntry entry)
IdentityMap`1.Add(InternalEntityEntry entry)
StateManager.StartTracking(InternalEntityEntry entry)
InternalEntityEntry.SetEntityState(EntityState oldState, EntityState newState, Boolean acceptChanges, Boolean modifyProperties)
InternalEntityEntry.SetEntityState(EntityState entityState, Boolean acceptChanges, Boolean modifyProperties, Nullable`1 forceStateWhenUnknownKey, Nullable`1 fallbackState)
EntityGraphAttacher.PaintAction(EntityEntryGraphNode`1 node)
EntityEntryGraphIterator.TraverseGraph[TState](EntityEntryGraphNode`1 node, Func`2 handleNode)
EntityGraphAttacher.AttachGraph(InternalEntityEntry rootEntry, EntityState targetState, EntityState storeGeneratedWithKeySetTargetState, Boolean forceStateWhenUnknownKey)
InternalDbSet`1.SetEntityState(InternalEntityEntry entry, EntityState entityState)
InternalDbSet`1.Add(TEntity entity)
RepositoryBase`1.AddAsync(T entity, CancellationToken cancellationToken)
EfRepositoryAdd.AddsProjectAndSetsId() line 19
--- End of stack trace from previous location ---

```

**Comment by SteveDunn** [comment]
Hi Steve, you were missing the `ProjectId` from the marker class:

![image](https://github.com/user-attachments/assets/e22a728e-d14c-4382-9e3b-b316c3b01a94)

This worked when creating projects, but failed when adding a todo item. This was because the specifications used the primitive type in the where clause. Changing this to `ProjectId` fixed it, and meant that `int`s needed to be changed `ProjectId` in other places. This was nice though as it meant fewer primitives in the core/domain and use-case layers.

I'll update my PR with the changes

**Reply by ardalis** [reply]
Ok that was dumb. Not sure why I only had ProjectName in there... thanks again!
